llvm/unittests/DebugInfo/LogicalView/DWARFGeneratedTest.cpp

```diff
@@ -0,0 +1,369 @@
+  EXPECT_EQ(TypeImplicitRangeDieDG.getTag(), DW_TAG_subrange_type);
+
+  // Save the generated DWARF file to disk.
+  EXPECT_TRUE(DG->saveFile(Path));
+}
+
+// Check the logical elements basic properties.
+void checkElementAttributes(LVReader *Reader) {
+  LVScopeRoot *Root = Reader->getScopesRoot();
+  EXPECT_NE(Root, nullptr);
+  LVScopeCompileUnit *CompileUnit = getFirstCompileUnit(Root);
+
+  const LVScopes *Scopes = CompileUnit->getScopes();
+  ASSERT_NE(Scopes, nullptr);
+  ASSERT_EQ(Scopes->size(), 3u);
+
+  // Check values.
+  LVScopes::const_iterator ScopeIter = Scopes->begin();
+  EXPECT_NE(ScopeIter, nullptr);
+  LVScope *Scope = static_cast<LVScope *>(*ScopeIter);
+  EXPECT_NE(Scope, nullptr);
+  EXPECT_EQ(Scope->getAccessibilityCode(), 1); // Element
+  EXPECT_EQ(Scope->getInlineCode(), 2);        // Element
+  EXPECT_EQ(Scope->getVirtualityCode(), 3);    // Element
+  EXPECT_EQ(Scope->getCallFilenameIndex(), 5); // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getCallLineNumber(), 5);    // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getFilenameIndex(), 7);     // Element
+  EXPECT_EQ(Scope->getLineNumber(), 7);        // Element
+  EXPECT_EQ(Scope->getDiscriminator(), 8);     // ScopeFunctionInlined
+
+  // Check no-values.
+  EXPECT_NE(++ScopeIter, nullptr);
+  Scope = static_cast<LVScope *>(*ScopeIter);
+  EXPECT_NE(Scope, nullptr);
+  EXPECT_EQ(Scope->getAccessibilityCode(), 0); // Element
+  EXPECT_EQ(Scope->getInlineCode(), 0);        // Element
+  EXPECT_EQ(Scope->getVirtualityCode(), 0);    // Element
+  EXPECT_EQ(Scope->getCallFilenameIndex(), 1); // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getCallLineNumber(), 0);    // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getFilenameIndex(), 1);     // Element
+  EXPECT_EQ(Scope->getLineNumber(), 0);        // Element
+  EXPECT_EQ(Scope->getDiscriminator(), 0);     // ScopeFunctionInlined
+
+  // Check implicit values.
+  EXPECT_NE(++ScopeIter, nullptr);
+  Scope = static_cast<LVScope *>(*ScopeIter);
+  EXPECT_NE(Scope, nullptr);
+  EXPECT_EQ(Scope->getAccessibilityCode(), 1); // Element
+  EXPECT_EQ(Scope->getInlineCode(), 2);        // Element
+  EXPECT_EQ(Scope->getVirtualityCode(), 3);    // Element
+  EXPECT_EQ(Scope->getCallFilenameIndex(), 5); // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getCallLineNumber(), 5);    // ScopeFunctionInlined
+  EXPECT_EQ(Scope->getFilenameIndex(), 7);     // Element
+  EXPECT_EQ(Scope->getLineNumber(), 7);        // Element
+  EXPECT_EQ(Scope->getDiscriminator(), 8);     // ScopeFunctionInlined
+
+  const LVSymbols *Symbols = CompileUnit->getSymbols();
+  ASSERT_NE(Symbols, nullptr);
+  ASSERT_EQ(Symbols->size(), 3u);
+
+  // Check values.
+  LVSymbols::const_iterator SymbolIter = Symbols->begin();
+  EXPECT_NE(SymbolIter, nullptr);
+  LVSymbol *Symbol = static_cast<LVSymbol *>(*SymbolIter);
+  EXPECT_NE(Symbol, nullptr);
+  EXPECT_EQ(Symbol->getBitSize(), 1); // Symbol
+
+  // Check no-values.
+  EXPECT_NE(++SymbolIter, nullptr);
```

Comment:
This pattern continues, I won't add comments on all of them.

Reply:
Removed the repeated comments.